include/pybind11/pytypes.h

```diff
@@ -930,11 +930,10 @@ class str : public object {
 private:
     /// Return string representation -- always returns a new reference, even if already a str
     static PyObject *raw_str(PyObject *op) {
-        PyObject *str_value = PyObject_Str(op);
-        if (!str_value) throw error_already_set();
 #if PY_MAJOR_VERSION < 3
-        PyObject *unicode = PyUnicode_FromEncodedObject(str_value, "utf-8", nullptr);
-        Py_XDECREF(str_value); str_value = unicode;
+        PyObject *str_value = PyObject_Unicode(op);
+#else
+        PyObject *str_value = PyObject_Str(op);
```

Comment:
We already have a ton of compatibility macros in `common.h`. I'd define a new one to avoid this macro branching here. Something like `PYBIND11_OBJECT_TO_STRING`.

Reply:
> We already have a ton of compatibility macros in `common.h`. I'd define a new one to avoid this macro branching here. Something like `PYBIND11_OBJECT_TO_STRING`.

Thanks Boris, I'm inclined to take your suggestions, although it makes it a (slightly) sprawling change.
I just looked: the needed compatibility macro doesn't exist already, I'll have to add it.
Checking with @YannickJadoul and @EricCousineau-TRI : what's your recommendation, local change as-is or adding the compatibility macro?

Reply:
I'd be fine with a compatibility macro living in `common.h`. The main concern about macros is some sort of deprecation mechanism when it's no longer necessary, but I'm sure there are several ways to resolve that.